pulsar-transaction/buffer/src/main/java/org/apache/pulsar/transaction/buffer/impl/PersistentTransactionBuffer.java

```diff
@@ -0,0 +1,308 @@
+import org.apache.pulsar.transaction.buffer.TransactionMeta;
+import org.apache.pulsar.transaction.buffer.exceptions.TransactionNotSealedException;
+import org.apache.pulsar.transaction.impl.common.TxnID;
+import org.apache.pulsar.transaction.impl.common.TxnStatus;
+
+/**
+ * A persistent transaction buffer implementation.
+ */
+@Slf4j
+public class PersistentTransactionBuffer extends PersistentTopic implements TransactionBuffer {
+
+    private TransactionCursor txnCursor;
+    private ManagedCursor retentionCursor;
+
+
+    abstract class TxnCtx implements PublishContext{
+        private long sequenceId;
+
+        TxnCtx(long sequenceId) {
+            this.sequenceId = sequenceId;
+        }
+
+        @Override
+        public String getProducerName() {
+            return "txn-producer";
+        }
+
+        @Override
+        public long getSequenceId() {
+            return this.sequenceId;
+        }
+    }
+
+    @Builder
+    private static final class Marker implements Serializable {
+        TxnID txnID;
+        TxnStatus status;
+
+        public byte[] serialize() {
+            return SerializationUtils.serialize(this);
+        }
+    }
+
+    public PersistentTransactionBuffer(String topic, ManagedLedger ledger, BrokerService brokerService)
+        throws BrokerServiceException.NamingException, ManagedLedgerException {
+        super(topic, ledger, brokerService);
+        this.txnCursor = new TransactionCursorImpl();
+        this.retentionCursor = ledger.newNonDurableCursor(PositionImpl.earliest);
+    }
+
+    @Override
+    public CompletableFuture<TransactionMeta> getTransactionMeta(TxnID txnID) {
+        return txnCursor.getTxnMeta(txnID, false);
+    }
+
+    @Override
+    public CompletableFuture<Void> appendBufferToTxn(TxnID txnId, long sequenceId, ByteBuf buffer) {
+        CompletableFuture<Void> appendFuture = new CompletableFuture<>();
+
+        publishMessage(buffer, new TxnCtx(sequenceId) {
+            @Override
+            public void completed(Exception e, long ledgerId, long entryId) {
+                if (e != null) {
+                    appendFuture.completeExceptionally(e);
+                    return;
+                }
+
+                txnCursor.getTxnMeta(txnId, true).thenCompose(meta -> {
+                    meta.appendEntry(sequenceId, PositionImpl.get(ledgerId, entryId));
+                    appendFuture.complete(null);
+                    return null;
+                }).exceptionally(metaNotFound -> {
+                    appendFuture.completeExceptionally(metaNotFound);
+                    return null;
+                });
+            }
+        });
+
+        return appendFuture;
+    }
+
+    @Override
+    public CompletableFuture<TransactionBufferReader> openTransactionBufferReader(TxnID txnID, long startSequenceId) {
+        CompletableFuture<TransactionBufferReader> readerFuture = new CompletableFuture<>();
+
+        txnCursor.getTxnMeta(txnID, false).thenCompose(meta -> {
```

Comment:
you don't need to use thenCompose. `thenApply` is enough.

I would suggest you change to use `whenComplete` as commented above.